README.md
jquery-simple-datetimepicker (jquery.simple-dtpicker.js)
========

Date and time picker, it's simple & clean. (jQuery plugin)

![screenshot](https://raw.github.com/mugifly/jquery-simple-datetimepicker/master/design/dtpicker_screenshot.png)

Details and examples: http://mugifly.github.com/jquery-simple-datetimepicker

Your feedback is highly appreciated.
You can send requests and bug reports as [Issues on GitHub](https://github.com/mugifly/jquery-simple-datetimepicker/issues).

## Requirements

This plugin requires these either version of jQuery.

* jQuery 1.7.2 or later
* jQuery 2.x

## For users

The master branch is development version.
Instead you can get a stable version from [released tags](https://github.com/mugifly/jquery-simple-datetimepicker/releases).

## For developers

An automated testing status of *master* branch: [![Build Status](https://travis-ci.org/mugifly/jquery-simple-datetimepicker.svg?branch=master)](https://travis-ci.org/mugifly/jquery-simple-datetimepicker)

An example page of latest *master* branch is [here](http://jq-simple-dtpicker-gh-master.herokuapp.com/jquery.simple-dtpicker.html);
or simply you can see an example page in your local.

### Testing on your environment

I don't force you to run a test of your commits. But recommends it.

https://github.com/mugifly/jquery-simple-datetimepicker/wiki#testing

### Contributors list

[List of contributors](https://github.com/mugifly/jquery-simple-datetimepicker/wiki/Contributors); Thanks :)

## License and maintainer

Copyright (c) 2013-2015 Masanori Ohgita (http://ohgita.info/).

This program is free software distributed under the terms of the MIT license.
See LICENSE file for details.


CHANGES
Version history for query-simple-datetimepicker
https://github.com/mugifly/jquery-simple-datetimepicker/

1.14.0  2018-10-29
  - Merge long waiting pull requests from many contributors, mainly:
    - Locales: icelandic, finnish.
    - [Add] Close button option.
    - CSS-Compatibility with JQuery 3.x
    - Replace deprecated .bind() calls with .on()
    - Correction and completion of the PT local, option to specify time range available in the list.
    - fix double rendering when navigating to previousMonth

1.13.0  2015-xx-xx (JST)
  - [Add] Locales: Ukrainian, Hungarian, Danish, Estonian, Norwegian.
  - [Add] Close button option.
  - [Add] minDate option and setMinDate function.
  - [Add] timeOnly option.
  - [Fix] isPast logic when calendar is in next month display.
  - [Fix] Locales: Spanish, Korean.

1.12.0  2014-04-25 (JST)
  - [Add] Automatically relocate a position of a picker when picker is float-mode.
  - [Add] Automatically destroy a picker, when an input-field has lose.
  - [Add] The getDate and setDate method is available on PickerHandler Object.
  - [Change] Changed a parent element to root, when picker is float-mode (Reported by: Kupid).

1.11.0  2014-04-25 (JST)
  - [Add] Support for allowWdays option (Idea by: elvinlee).
  - [Add] Greek locale (Thanks to: sincetomorrow).
  - [Fix] Issue at option.current - Setting the custom date and time as default (Thanks to: michelts).
  - [Fix] Process for drawing the time cell (Reported by: jhiver).
  - [Fix] Past date is selected when clicking on previous (futureOnly) (Reported by: ignatiusreza).
  - [Fix] Bug when switching to next month.
  - [Fix] If not available a console on this environment.
  - [Fix] Position / outerHeight() gives wrong value (Thanks to: ricardocanelas).

1.10.1  2014-02-16 (JST)
  - [Fix] Issue with month jumping (Report by: jcutrell, LinTicket, cnorthwood).

1.10.0  2014-02-16 (JST)
  - [Add] Polish locale (Thanks to: viszman).
  - [Add] Romanian locale (Thanks to: IngerAlHaosului).
  - [Add] Support for minDate and maxDate options (Thanks to: droidmunkey).
  - [Add] Support for minTime and maxTime options (Thanks to: CedricLevasseur).
  - [Add] Support for user event-handlers: onInit, onShow, onHide (Idea by: Kupid Wang).
  - [Add] Manual handling of the appended picker (handleDtpicker method).
  - [Add] Allow focus of input to display datepicker (Thanks to: jpetto).
  - [Fix] Use isNaN instead of Number.isNaN (Thanks to: vbatoufflet).
  - [Fix] Don't allow scroll to past-month when limiting to future dates (Report by: fredrik-w).

1.9.0  2013-11-06 (JST)
  - [Add] Support for Bower (Thanks to: kmulvey).
  - [Change] Improved localization mechanism (Thanks to: LucasTheCure).
  - [Change] Refactor - Simplified some codes (Thanks to: LucasTheCure).
  - [Fix] Close a picker when date is selected on dateOnly mode (Thanks to: Kwozzie).

1.8.0  2013-10-17 (JST)
  - [Add] Portuguese locale (Thanks to: adc-tcarvalho).
  - [Add] Dutch locale (Thanks to: Mondane).
  - [Add] Czech locale (Thanks to: Jan Zatloukal).
  - [Add] French locale (Thanks to: Jan matll42).
  - [Add] Italian locale (Thanks to: Mauro Zuccato).
  - [Fix] Fixing positioning when element is inside of nested relative parent (Thanks to: AddoSolutions).

1.7.0  2013-09-16
  - [Add] Korean locale (Thanks to: perhapsspy).
  - [Add] "futureOnly" option and bugfix (Thanks to: camiel).
  - [Add] "dateOnly" option (Thanks to: logitick).
  - [Fix] Improved date parse (when using dateFormat option).

1.6.0  2013-09-04
  - [Add] "closeOnSelected" option for close the picker when selected date & time (Idea by: MacDknife).
  - [Add] "timelistScroll" option for disable automatic scroll of timelist (Idea by: deviso).
  - [Add] Change the months by scroll of mouse-wheel on the calendar (Idea by: deviso).
  - [Add] "calendarMouseScroll" option for scroll calendar with mouse wheel (default: true).
  - [Add] Today button for calendar (Idea by: Tanver186).
  - [Add] "todayButton" option (default: true).
  - [Change] Improve design (header, container).
  - [Fix] Malfunction of "animation" option (Can't disable animation).

1.5.0  2013-08-28
  - [Add] Turkish locale (Thanks to: myfell).
  - [Add] Spanish locale (Thanks to: maw).
  - [Add] Swedish locale (Thanks to: MacDknife).
  - [Fix] Be more robust when receiving clearly bogus date strings (Thanks to: maw).

1.4.0  2013-08-02
  - [Add] Indonesia locale (Thanks to: robzlabz).
  - [Fix] "firstDayOfWeek" option and others (Thanks to: thisispiers).
  - [Fix] should now gets format mysql-datetime (Thanks to: Orlac).

1.3.0  2013-06-29
  - [Add] Chinese locale (Thanks to: jasonslyvia).
  - [Add] Initial first day of the week (Thanks to: concept1hundred).
  - [Add] German locale (Thanks to: rtakeda).
  - [Change] Option name: "minute_interval" -> "minuteInterval"
  - [Fix] misplaced when located in nested elements (Thanks to: insonifi).

1.2.0  2013-06-17
  - [Add] Change the interval of minute (Thanks to: AndAniCalik).

1.1.4  2013-06-02
  - [Add] Brasil locale (Thanks to: Mauricio).

1.1.3  2013-05-22
  - [Add] Russian locale (Thanks to: rdolgushin).

1.1.2  2013-05-09
  - [Fixed] IE8 compatibility (Thanks to: mackeian).

1.1.1  2013-03-12
  - [Fixed] Fixed width of timelist for bug in ie9 (Thanks to: jjsgro).

1.1.0  2013-03-02
   - [Implement] Option to disable the animation.
   - [Improve] Animation and parts.
   - [Fixed] Support dateFormat and fix bug (Thanks to: LeandroSe).
   - [Fixed] Support Firefox in current date (Thanks: LeandroSe).
   - [Improve] Show the month name instead of number in header (Thanks to: dsims).
   - [Fixed] appendDtpicker is not overriding the default date with the input's value (Thanks: dsims).
   - [Fixed] default month (Thanks: dsims).

1.0.0  2012-11-02
   - Initial release.

----

Thanks to contributors!


t/test.js
/**
	Test script.
	The date-picker elements is cleared when each test has finished.

	Based on QUnit: http://api.qunitjs.com/category/assert/

**/

$(function() {


	/* Basic test
		Picker is inline and standalone mode (That not append with an input-field). */
	test('Basic', function(){
		var $picker = $('#inline_date_picker');
		$picker.dtpicker();

		equal($picker.css('display'), "block");

		/* Calendar - Wdays */
		var $table = $('.datepicker_table');
		var $tr_wdays = $($table.children('tbody').children('tr')[0]);
		var $wday_cells = $tr_wdays.children('th');
		var wdays = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];
		for (var i = 0; i < 7; i++) {
			equal($($wday_cells[i]).text(), wdays[i]);
		}
	});


	/* Switching a month - Previous
		2014-01-01 00:00 -> 2013-12-01 00:00
		Picker is inline mode, and append into an input-field. */
	asyncTest('Switching a month - Previous', function(){
		expect(2);

		var $date_input = $('#date_input');

		// The field value should be initial date
		$date_input.val('2014-01-01 00:00');
		$date_input.appendDtpicker({
			'inline': true
		});

		var $picker_header = $('.datepicker_header');

		// Switch a month to previous
		var $prev_btn = $($picker_header.children('a')[1]);
		$prev_btn.click();

		setTimeout(function(){
			// The field value should be same with first
			equal($date_input.val(), '2014-01-01 00:00');

			// Choose a first day
			var $days = $('.datepicker_table td');
			var $first_day = null;
			$days.each(function() {
				if ($(this).text() == 1) {
					$first_day = $(this);
					return false;
				}
			});
			$first_day.click();

			setTimeout(function() {
				// The field value should be first day of previous month
				equal($date_input.val(), '2013-12-01 00:00');
				// Destroy a picker on the root element
				$date_input.handleDtpicker('destroy');
				// Done
				start();
			}, 100);

		}, 100);
	});


	/* Switching a month - Next
		2014-01-01 00:00 -> 2014-02-01 00:00
		Picker is inline mode, and append into an input-field. */
	asyncTest('Switching a month - Next', function(){
		expect(2);

		var $date_input = $('#date_input');

		// The field value should be initial date
		$date_input.val('2014-01-01 00:00');
		$date_input.appendDtpicker({
			'inline': true
		});

		var $picker_header = $('.datepicker_header');

		// Switch a month to next
		var $next_btn = $($picker_header.children('a')[2]);
		$next_btn.click();

		setTimeout(function(){
			// The field value should be same with first
			equal($date_input.val(), '2014-01-01 00:00');

			// Choose a first day
			var $days = $('.datepicker_table td');
			var $first_day = null;
			$days.each(function() {
				if ($(this).text() == 1) {
					$first_day = $(this);
					return false;
				}
			});
			$first_day.click();

			setTimeout(function() {
				// The field value should be first day of next month
				equal($date_input.val(), '2014-02-01 00:00');
				// Destroy a picker on the root element
				$date_input.handleDtpicker('destroy');
				// Done
				start();
			}, 100);

		}, 100);
	});


	/* handleDtpicker method
		Picker is inline mode, and append into an input-field. */
	test('handleDtpicker method', function(){
		var $date_input = $('#date_input');
		$date_input.val('2014-10-31 00:00'); // Initial date
		$date_input.appendDtpicker({
			'inline': true
		});
		var $picker = $('.datepicker');

		$date_input.handleDtpicker('show');
		equal($picker.css('display'), "block");

		$date_input.handleDtpicker('hide');
		equal($picker.css('display'), "none");

		$date_input.handleDtpicker('setDate', new Date(2014, 11, 01, 12, 10, 0));

		// Check a date
		var date = $date_input.handleDtpicker('getDate');
		equal(date.getFullYear(), 2014);
		equal(date.getMonth(), 11);
		equal(date.getDate(), 1);
		equal(date.getHours(), 12);
		equal(date.getMinutes(), 10);
	});


	/* Automatically destroy
		Picker is float mode, and append into an input-field. */
	asyncTest('Automatically destroy', function(){
		expect(1);
		var $date_input = $('#date_input');
		$date_input.val('2014-10-31 00:00'); // Initial date
		$date_input.appendDtpicker({
			'inline': false
		});

		// Delete an input-field
		$date_input.remove();
		setTimeout(function(){
			var $picker = $('.datepicker');
			equal($picker.length, 0); // There is not a single picker in it.
			start(); // Done.
		}, 100);
	});


	/* Option - Inline : true (Inline mode)
		Picker is inline mode, and append into an input-field. */
	test('Option - inline : true (Inline mode)', function(){
		var $date_input = $('#date_input');
		$('#date_input').appendDtpicker({
			'inline' : true
		});

		var $picker = $('.datepicker');
		equal($picker.css('display'), "block");
	});


	/* Option - Inline : FALSE (Float mode)
		Picker is float mode, and append into an input-field.
		In this test, Picker has shown when click and focus a field. */
	test('Option - inline : false (Float mode)', function(){
		var $date_input = $('#date_input');
		$date_input.appendDtpicker({
			'inline': false
		});

		var $picker = $('.datepicker');
		equal($picker.css('display'), "none");

		var $date_input = $('#date_input');
		$date_input.text('2014/04/24 00:00');
		$date_input.bind('focus', function(){
			var $picker = $('.datepicker');
			equal($picker.css('display'), "block");
			$date_input.unbind('focus');

			// Destroy a picker on the root element
			$date_input.handleDtpicker('destroy');
		});
		$date_input.triggerHandler('click');
		$date_input.triggerHandler('focus');
	});


	/* Option - minuteInterval */
	test('Option - minuteInterval', function(){
		var $date_input = $('#date_input');
		$date_input.appendDtpicker({
			'inline': true,
			'minuteInterval': 15
		});

		var $timelist = $('.datepicker_timelist');
		var $time_cells = $timelist.children('div');
		equal($($time_cells[0]).text(), '00:00');
		equal($($time_cells[1]).text(), '00:15');
		equal($($time_cells[$time_cells.length  - 1]).text(), '23:45');
	});

	/* Option - firstDayOfWeek (Change the first day of the week) */
	test('Option - firstDayOfWeek', function(){
		var $date_input = $('#date_input');
		$date_input.appendDtpicker({
			'inline': true,
			'firstDayOfWeek': 1
		});

		var $table = $('.datepicker_table');
		var $tr_wdays = $($table.children('tbody').children('tr')[0]);
		var $wday_cells = $tr_wdays.children('th');
		equal($($wday_cells[0]).text(), 'Mo'); // Monday
		equal($($wday_cells[6]).text(), 'Su'); // Sunday
	});


	/* Option - minTime & maxTime */
	asyncTest('Option - minTime & maxTime', function(){

		// Initialize
		var $date_input = $('#date_input');
		$date_input.val('2015-01-01 00:00:00');
		$date_input.appendDtpicker({
			inline: true,
			minTime: '08:30',
			maxTime: '19:15'
		});

		// Check the time list
		var chkTimeList = function() {

			var $picker_time_list = $('.datepicker_timelist');
			var $times = $picker_time_list.children('.timelist_item');

			$times.each(function(){
				var $time = $(this);
				var t = $time.text().split(/:/);
				if ((t[0] == 8 && 30 <= t[1]) || (t[0] == 19 && t[1] <= 15) || (9 <= t[0] && t[0] <= 18)) { // It expect as valid time
					equal($time.hasClass('time_in_past'), false, $time.text() + ' is valid time');
				} else {
					assert.notOk(false, $time.text() + ' is invalid time. It should not shown.');
				}
			});

		};
		chkTimeList();

		// Switch to next month
		var $picker_header = $('.datepicker_header');
		var $next_button = $($picker_header.children('a')[2]);
		$next_button.click();
		setTimeout(function(){

			// Check the time list again
			chkTimeList();

			// Test was done
			$date_input.handleDtpicker('destroy'); // Destroy a picker on the root element
			start();

		}, 100);

	});


});


Gruntfile.js
module.exports = function(grunt) {
	grunt.initConfig({
		pkg: '<json:package.json>',
		// Test with using QUnit
		qunit: {
			files: ['t/test.html']
		},
		// Syntax check with using JSHint
		jshint: {
			default: {
				src: ['jquery.simple-dtpicker.js'],
				options: {
					eqnull: true,
					loopfunc: true,
				},
				predef: [
					'eqnull', 'eqeqeq'
				]
			}
		},
		// Sync manifest for bower.json
		manifestSync: {
			dist: {
				options: {
					primaryManifest: "package.json",
					manifests: {
						bower: "bower.json"
					}
				}
			}
		}
	});

	// Load a module
	grunt.loadNpmTasks('grunt-contrib-qunit');
	grunt.loadNpmTasks('grunt-contrib-jshint');
	grunt.loadNpmTasks('grunt-manifest-sync');

	// Register a tasks
	grunt.registerTask('test', ['qunit', 'jshint']);
	grunt.registerTask('default', ['manifestSync', 'test']);
}


dtpicker.js
/**
 * dtpicker (jquery-simple-datetimepicker)
 * (c) Masanori Ohgita - 2012.
 * https://github.com/mugifly/jquery-simple-datetimepicker
 */

(function($) {
	var DAYS_OF_WEEK_EN = ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'];
	var DAYS_OF_WEEK_JA = ['日', '月', '火', '水', '木', '金', '土'];

	var pickerObjects = [];
	var inputObjects = [];

	var getParentPickerObject = function(obj) {
		var $obj = $(obj);
		var $picker;
		if ($obj.hasClass('datepicker')) {
			$picker = $obj;
		} else {
			var parents = $obj.parents();
			for (var i = 0; i < parents.length; i++) {
				if ($(parents[i]).hasClass('datepicker')) {
					$picker = $(parents[i]);
				}
			}
			console.log($picker);
		}
		return $picker;
	};

	var getPickersInputObject = function($obj) {
		var $picker = getParentPickerObject($obj);
		console.log($picker.data());
		if ($picker.data("inputObjectId") != null) {
			return $(inputObjects[$picker.data("inputObjectId")]);
		}
		return null;
	}
	var beforeMonth = function($obj) {
		var $picker = getParentPickerObject($obj);
		var date = getPickedDate($picker);
		var targetMonth_lastDay = new Date(date.getYear() + 1900, date.getMonth(), 0).getDate();
		console.log(targetMonth_lastDay);
		if (targetMonth_lastDay < date.getDate()) {
			date.setDate(targetMonth_lastDay);
		}
		draw($picker, true, true, date.getYear() + 1900, date.getMonth() - 1, date.getDate(), date.getHours(), date.getMinutes());
	};

	var nextMonth = function($obj) {
		var $picker = getParentPickerObject($obj);
		var date = getPickedDate($picker);
		var targetMonth_lastDay = new Date(date.getYear() + 1900, date.getMonth() + 1, 0).getDate();
		if (targetMonth_lastDay < date.getDate()) {
			date.setDate(targetMonth_lastDay);
		}
		draw($picker, true, true, date.getYear() + 1900, date.getMonth() + 1, date.getDate(), date.getHours(), date.getMinutes());
	};

	var outputToInputObject = function($obj) {
		var date = getPickedDate($obj);
		var $inp = getPickersInputObject($obj);
		if ($inp != null) {
			$inp.val(date.getYear() + 1900 + "/" + zpadding(date.getMonth() + 1) + "/" + zpadding(date.getDate()) + "  " + zpadding(date.getHours()) + ":" + zpadding(date.getMinutes()));
		}
	};

	var getPickedDate = function($obj) {
		var $picker = getParentPickerObject($obj);
		return $picker.data("pickedDate");
	};

	var zpadding = function(num) {
		num = ("0" + num).slice(-2);
		return num
	};

	var draw_date = function($picker, isAnim, isOutputToInputObject, date) {
		draw($picker, isAnim, isOutputToInputObject, date.getYear() + 1900, date.getMonth(), date.getDate(), date.getHours(), date.getMinutes());
	};

	var draw = function($picker, isAnim, isOutputToInputObject, year, month, day, hour, min) {
		var date = new Date();
		
		if (hour != null) {
			date = new Date(year, month, day, hour, min, 0);
		} else if (year != null) {
			date = new Date(year, month, day);
		} else {
			date = new Date();
		}
		console.log("dtpicker - draw()..." + year + "," + month + "," + day + " " + hour + ":" + min + " -> " + date);

		$($picker).data("pickedDate", date);

		var todayDate = new Date(); 
		var firstWday = new Date(date.getYear() + 1900, date.getMonth(), 1).getDay();
		var lastDay = new Date(date.getYear() + 1900, date.getMonth() + 1, 0).getDate();
		var beforeMonthLastDay = new Date(date.getYear() + 1900, date.getMonth(), 0).getDate();
		var dateBeforeMonth = new Date(date.getYear() + 1900, date.getMonth(), 0);
		var dateNextMonth = new Date(date.getYear() + 1900, date.getMonth() + 2, 0);

		var $inner = $picker.children('.datepicker_inner_container');

		if (isAnim == true) {
			$inner.stop().queue([]);
			$inner.fadeTo("fast", 0.5);
		}

		/* Header */
		var $header = $picker.children('.datepicker_header');
		$header.children().remove();
		var $link_before_month = $('<a>');
		$link_before_month.text('<');
		$link_before_month.click(function() {
			beforeMonth($picker);
		});

		var $now_month = $('<span>');
		$now_month.text((date.getYear() + 1900) + " / " + zpadding(date.getMonth() + 1));

		var $link_next_month = $('<a>');
		$link_next_month.text('>');
		$link_next_month.click(function() {
			nextMonth($picker);
		});

		$header.append($link_before_month);
		$header.append($now_month);
		$header.append($link_next_month);

		/* Calendar > Table */
		var $calendar = $picker.children('.datepicker_inner_container').children('.datepicker_calendar');
		var $table = $calendar.children('.datepicker_table');
		$table.children().remove();
		var $tr = $('<tr>');
		$table.append($tr);

		/* Output wday cells */
		for (var i = 0; i < 7; i++) {
			var $td = $('<th>');
			$td.text(DAYS_OF_WEEK_EN[i]);
			$tr.append($td);
		}

		/* Output day cells */
		var cellNum = Math.ceil((firstWday + lastDay) / 7) * 7;
		for (var i = 0; i < cellNum; i++) {
			var realDay = i + 1 - firstWday;
			if (i % 7 == 0) {
				$tr = $('<tr>');
				$table.append($tr);
			}

			var $td = $('<td>');
			$td.data("day", realDay);
			
			$tr.append($td);
			
			if (firstWday > i) {/* Before months day */
				$td.text(beforeMonthLastDay + realDay);
				$td.addClass('day_another_month');
				$td.data("dateStr", dateBeforeMonth.getYear() + 1900 + "/" + (dateBeforeMonth.getMonth() + 1) + "/" + (beforeMonthLastDay + realDay));
			} else if (i < firstWday + lastDay) {/* Now months day */
				$td.text(realDay);
				$td.data("dateStr", (date.getYear() + 1900) + "/" + (date.getMonth() + 1) + "/" + realDay);
			} else {/* Next months day */
				$td.text(realDay - lastDay);
				$td.addClass('day_another_month');
				$td.data("dateStr", dateNextMonth.getYear() + 1900 + "/" + (dateNextMonth.getMonth() + 1) + "/" + (realDay - lastDay));
			}

			if (i % 7 == 0) {/* Sunday */
				$td.addClass('wday_sun');
			} else if (i % 7 == 6) {/* Saturday */
				$td.addClass('wday_sat');
			}

			if (realDay == date.getDate()) {/* selected day */
				$td.addClass('active');
			}
			
			if (date.getMonth() == todayDate.getMonth() && realDay == todayDate.getDate()) {/* today */
				$td.addClass('today');
			}

			/* Set event-handler to day cell */

			$td.click(function() {
				if ($(this).hasClass('hover')) {
					$(this).removeClass('hover');
				}
				$(this).addClass('active');

				var $picker = getParentPickerObject($(this));
				console.log($(this).data("dateStr"));
				var targetDate = new Date($(this).data("dateStr"));
				var selectedDate = getPickedDate($picker);
				draw($picker, false, true, targetDate.getYear() + 1900, targetDate.getMonth(), targetDate.getDate(), selectedDate.getHours(), selectedDate.getMinutes());
			});

			$td.hover(function() {
				if (! $(this).hasClass('active')) {
					$(this).addClass('hover');
				}
			}, function() {
				if ($(this).hasClass('hover')) {
					$(this).removeClass('hover');
				}
			});
		}

		/* Timelist */
		var $timelist = $picker.children('.datepicker_inner_container').children('.datepicker_timelist');
		$timelist.children().remove();
		
		/* Set height to Timelist (Calendar innerHeight - Calendar padding) */
		$timelist.css("height", $calendar.innerHeight() - 10 + 'px');

		/* Output time cells */
		for (var hour = 0; hour < 24; hour++) {
			for (var min = 0; min <= 30; min += 30) {
				var $o = $('<div>');
				$o.addClass('timelist_item');
				$o.text(zpadding(hour) + ":" + zpadding(min));

				$o.data("hour", hour);
				$o.data("min", min);

				$timelist.append($o);

				if (hour == date.getHours() && min == date.getMinutes()) {/* selected time */
					$o.addClass('active');
				}

				/* Set event handler to time cell */
				
				$o.click(function() {
					if ($(this).hasClass('hover')) {
						$(this).removeClass('hover');
					}
					$(this).addClass('active');

					var $picker = getParentPickerObject($(this));
					var date = getPickedDate($picker);
					var hour = $(this).data("hour");
					var min = $(this).data("min");
					draw($picker, false, true, date.getYear() + 1900, date.getMonth(), date.getDate(), hour, min);
				});
				
				$o.hover(function() {
					if (! $(this).hasClass('active')) {
						$(this).addClass('hover');
					}
				}, function() {
					if ($(this).hasClass('hover')) {
						$(this).removeClass('hover');
					}
				});
			}
		}

		if (isAnim == true) {
			$inner.fadeTo("fast", 1.0);
		}

		/* Output to InputForm */
		if (isOutputToInputObject == true) {
			outputToInputObject($picker);
		}
	};

	var init = function($obj, opt_inputObjectId) {
		console.log("dtpicker init... ");

		/* Container */
		var $picker = $('<div>');
		$picker.addClass('datepicker')
		$obj.append($picker);

		if (opt_inputObjectId != null) {
			$picker.data("inputObjectId", opt_inputObjectId);
		}
		$picker.data("pickerId", pickerObjects.length);

		/* Header */
		var $header = $('<div>');
		$header.addClass('datepicker_header');
		$picker.append($header);
		/* InnerContainer*/
		var $inner = $('<div>');
		$inner.addClass('datepicker_inner_container');
		$picker.append($inner);
		/* Calendar */
		var $calendar = $('<div>');
		$calendar.addClass('datepicker_calendar');
		var $table = $('<table>');
		$table.addClass('datepicker_table');
		$calendar.append($table);
		$inner.append($calendar);
		/* Timelist */
		var $timelist = $('<div>');
		$timelist.addClass('datepicker_timelist');
		$inner.append($timelist);

		pickerObjects.push($picker);

		draw($picker, true, true);
	};

	/* Initialize dtpicker */
	$.fn.dtpicker = function(config) {
		var defaults = {

		}
		var options = $.extend(defaults, config);
		return this.each(function(i) {
			if(config == null || config.inputObjectId == undefined){
				init($(this));
			}else{
				init($(this), config.inputObjectId);
			}
		});
	};

	/* Initialize dtpicker, append to Text input field */
	$.fn.appendDtpicker = function(config) {
		var defaults = {
			"inline": false
		}
		var options = $.extend(defaults, config);
		return this.each(function(i) {
			/* Add input-field with inputsObjects array */
			var input = this;
			var inputObjectId = inputObjects.length;
			inputObjects.push(input);

			/* Make parent-div for picker */
			console.log(this);
			var $d = $('<div>');
			$d.insertAfter(input);

			var pickerId = pickerObjects.length;

			/* Initialize picker */
			var $picker_parent = $($d).dtpicker({
				"inputObjectId" : inputObjectId
			});

			/* Link input-field with picker*/
			$(input).data('pickerId', pickerId);
			
			/* Set event handler to input-field */
			
			$(input).keyup(function() {
				var $input = $(this);
				var $picker = $(pickerObjects[$input.data('pickerId')]);
				if ($input.val() != null) {
					var date = new Date($input.val());
					if (isNaN(date.getDate()) == false) {/* Valid format... */
						console.log(date.getDate())
						draw_date($picker, true, false, date);
					}
				}
			});
			
			$(input).change(function(){
				$(this).trigger('keyup');
			});
			
			/* for NOT inline mode */
			if(options.inline == false){
				$picker_parent.css("position","absolute");
				$picker_parent.hide();
				$(input).click(function(){
					var $input = $(this);
					var $picker = $(pickerObjects[$input.data('pickerId')]);
					$($picker.parent()).show();
				});
			}

		});
	};
})(jQuery); 